### PR TITLE
feat: add temporal stability validation metrics

### LINKED ISSUE #181: Add temporal stability validation metrics
## Background

The current `tier4-main` validation metrics cover planning accuracy, collision-related scores, progress, feasibility, centerline alignment, and comfort-like smoothness. They do not explicitly measure temporal stability: whether the planner output remains smooth within one trajectory and whether replanning on adjacent frames produces consistent overlapping future trajectories.

This matters because open-loop trajectory loss can improve while the planner still flickers frame-to-frame. For onboard planning and closed-loop validation, that flicker can become unstable steering or unstable downstream post-processing even when single-frame displacement errors look acceptable.

### 1. Single-trajectory stability probes

Add validation subscores derived from the predicted ego trajectory:

- `jerk` / smoothness: estimate trajectory jerk from the predicted XY path and report a lower-is-better physical smoothness signal.
- `curvature_rate`: estimate curvature change rate along the predicted trajectory. This catches steering oscillation that may not be obvious from position error alone.

The implementation should avoid raw high-order finite differences where possible because our trajectories are sampled at 10 Hz and raw derivatives amplify noise. A smoothed derivative implementation, such as a Savitzky-Golay style convolution, is preferred for stable metric values.

### 2. Cross-frame replan consistency

Add an open-loop temporal consistency metric for consecutive NPZ frames:

1. Take the prediction generated at frame `t` and the prediction generated at frame `t + g`.
2. Use ego ground-truth motion between those frames to express the frame-`t` prediction in the frame-`t + g` ego coordinate system.
3. Compare the overlapping future horizon between the transformed earlier prediction and the later prediction.
4. Report mean position jump and heading jump over the overlap.

Lower values mean the planner is more temporally stable. This is an open-loop proxy for closed-loop replan flicker: if two adjacent replans describe the same wall-clock future, their overlapping portions should agree unless the scene genuinely changes.

## Scope

This issue is for validation metrics only. It should not introduce a training loss, model architecture change, JEPA/prior-conditioning path, or DFP-specific behavior.

The metrics should be usable for:

- original Diffusion Planner baseline validation,
- SFT validation,
- DFP validation,
- full-sequence datasets where consecutive NPZ frames are available.

## Expected implementation requirements

- Add the metrics under `planner_metrics` with a small, documented API.
- Keep the default training behavior unchanged.
- Add unit tests for geometry, heading wrapping, overlap handling, grouping of consecutive frame paths, and differentiability where relevant.
- Keep the PR independent from DFP and ePDMS changes.
- Run local lint/pre-commit and targeted tests before marking the PR ready.

### PR DESCRIPTION
**Pull request by Max-Bin**
## Summary

Adds temporal-stability validation metrics:

- `Inter-frame Stability`: `replan_position_consistency` / `replan_heading_consistency` for adjacent Step-1 NPZ frames.
- `Mean Absolute Jerk`: `ego_mean_abs_jerk` for single-trajectory smoothness.
- `Curvature Rate`: `ego_curvature_rate` for steering oscillation.
- `DiffusionPlannerPairData`: a validation-only pair dataset that builds `(frame_t, frame_t+1)` pairs from the validation datalist when adjacent frames are available.

Closes #181.

## Usage

Temporal stability validation is enabled by default in this PR:

- `enable_temporal_stability_eval=True`
- `enable_replan_consistency_eval=True`
- `replan_consistency_expected_gap=1`

`ego_mean_abs_jerk` and `ego_curvature_rate` are single-trajectory metrics and can be computed from any validation datalist.

`Inter-frame Stability` requires a Step-1 full-sequence validation datalist because it compares adjacent replans. A normal skip-N datalist should not be used for this metric. With the default `replan_consistency_expected_gap=1`, skip-N lists produce no adjacent pairs, so the inter-frame metric is skipped instead of reporting a misleading value.

For experiments that still train on a skip-N datalist, use a separate Step-1 datalist for `--valid_set_list` if inter-frame consistency is part of the comparison.

## Why this needs pair loading

The normal Diffusion Planner validation dataloader samples one NPZ frame at a time, so it can evaluate single-frame trajectory quality but cannot evaluate frame-to-frame replanning stability. True `Inter-frame Stability` needs two adjacent frames that describe overlapping wall-clock future time.

This PR keeps the existing single-frame dataloader unchanged and adds a second validation-only pair dataloader for the inter-frame metric.

## Metric design

This PR logs the objective as validation metrics, not as a training loss. However, the calculations are intentionally loss-like and can be reused later as differentiable consistency losses.

### 1. Inter-frame Stability

For each adjacent pair:

- frame `a = t`
- frame `b = t + g`
- default `g = 1`, matching Step-1 full-sequence validation data
- `P_a[k] = (x, y, cos(yaw), sin(yaw))` is the model prediction made at frame `a`
- `P_b[k]` is the model prediction made at frame `b`
- `G_a[g-1]` is the ego GT future pose at frame `a` after `g` timesteps

`G_a[g-1]` gives the pose of the frame-`b` ego origin expressed in frame-`a` coordinates:

- translation `r = (r_x, r_y)`
- yaw `psi = atan2(sin, cos)`

The prediction made at frame `a` is transformed into frame `b` coordinates before comparison:

```text
xy_a_to_b[k] = R(-psi) * (P_a[g + k].xy - r)
yaw_a_to_b[k] = wrap(yaw(P_a[g + k]) - psi)
```

Then it is compared against `P_b[k]` over the overlapping horizon:

```text
L = min(T - g, T)
replan_position_consistency = mean_{k=0..L-1} || xy_a_to_b[k] - P_b[k].xy ||_2
replan_heading_consistency  = mean_{k=0..L-1} | wrap(yaw_a_to_b[k] - yaw(P_b[k])) |
```

Lower is better. If two adjacent replans describe the same future consistently, both values should be near zero. If the planner flickers laterally by 1m between frames, `replan_position_consistency` reports approximately 1m.

### 2. Mean Absolute Jerk

`ego_mean_abs_jerk` is computed from the predicted XY trajectory using the shared Savitzky-Golay derivative kernel from `planner_metrics.geometry`:

```text
jerk = d^3 xy / dt^3
ego_mean_abs_jerk = mean_t || jerk_t ||_2
```

Lower is smoother.

### 3. Curvature Rate

`ego_curvature_rate` estimates steering oscillation from smoothed derivatives:

```text
kappa = (v_x * a_y - v_y * a_x) / ||v||^3
ego_curvature_rate = mean_t | d kappa / dt |
```

Lower means less steering oscillation.

## Implementation notes

- Reuses `planner_metrics.geometry._build_sg_diff_kernel` instead of duplicating the SG kernel builder.
- Caches temporal-stability SG kernels by `(window, poly, deriv, dt, device, dtype)`.
- Shares validation input preparation between normal validation and pair-validation paths to avoid drift.
- Default `replan_consistency_expected_gap=1` prevents skip-N datalists from being silently treated as true adjacent replanning data.

## DDP aggregation

For `Inter-frame Stability`, each rank accumulates local sums:

```text
sum(replan_position_consistency), sum(replan_heading_consistency), sum(overlap_len), count
```

The sums are all-reduced across ranks, then divided by `count`. This avoids gathering all predictions on rank 0.

## Compatibility

- No model architecture, inference behavior, checkpoint format, optimizer behavior, or training loss is changed.
- The pair dataloader is validation-only.
- Single-frame validation still runs on ordinary datalists.
- Inter-frame consistency is only reported when adjacent Step-1 pairs are present.

## Tests

- filename parsing groups frames by scenario and frame index
- pair building keeps the modal consecutive gap and skips sequence breaks
- replan consistency is zero when overlapping predictions match after coordinate alignment
- a 1m lateral replan flicker reports approximately 1m position consistency
- jerk and curvature-rate are zero for a constant-speed straight trajectory

## Local validation

- Targeted metric tests passed: `5 passed`
- Targeted ruff passed
- Targeted pre-commit passed
- `git diff --check` passed

<details>
<summary>Commands</summary>

```bash
git diff --check
UV_CACHE_DIR=/tmp/uv-cache uv run ruff check planner_metrics/temporal_stability.py planner_metrics/test_temporal_stability_metrics.py diffusion_planner/diffusion_planner/validate_model.py diffusion_planner/diffusion_planner/train_config.py diffusion_planner/diffusion_planner/valid_config.py diffusion_planner/train_predictor.py diffusion_planner/valid_predictor.py
UV_CACHE_DIR=/tmp/uv-cache uv run python -m pytest planner_metrics/test_temporal_stability_metrics.py -q
UV_CACHE_DIR=/tmp/uv-cache PRE_COMMIT_HOME=/tmp/pre-commit-cache uv run pre-commit run --files planner_metrics/temporal_stability.py planner_metrics/test_temporal_stability_metrics.py diffusion_planner/diffusion_planner/validate_model.py diffusion_planner/diffusion_planner/train_config.py diffusion_planner/diffusion_planner/valid_config.py diffusion_planner/train_predictor.py diffusion_planner/valid_predictor.py docs/temporal_stability_validation.md
```

</details>

### CODE REVIEW

**Comment by SakodaShintaro** [comment]
conflicts!

**Comment by Max-Bin** [comment]
@SakodaShintaro 
Conflicts seems to have been caused by the PDMS merge. Solved now.

File: planner_metrics/temporal_stability.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    "compute_mean_abs_jerk_batch",
+    "compute_replan_consistency_batch",
+    "consecutive_frame_pairs",
+    "group_frames_by_scenario",
+    "inter_frame_transform",
+    "parse_frame_key",
+]
+
+
+def _build_sg_diff_kernel(window: int, poly: int, deriv: int, delta: float, device) -> torch.Tensor:
```
Comment (go-sakayori) [inline planner_metrics/temporal_stability.py]: This re-implements the SG kernel builder already exported from `planner_metrics/geometry.py` (`_build_sg_diff_kernel`), and `compute_mean_abs_jerk_batch` is `compute_smoothness_score_batch` (subscores.py) up to sign. Two concerns:

- The copy drops geometry's `.flip(0)` (conv1d convention). It's currently harmless since all outputs go through `norm`/`abs`, but a future fix to the shared kernel won't propagate to this third copy.
- Unlike `subscores.py`, the kernel isn't cached, so up to 3 `torch.linalg.pinv` solves run per validation batch.

Could this import the existing builder from `geometry.py` and cache the kernel by `(device, dtype)` like subscores does?

File: diffusion_planner/diffusion_planner/validate_model.py [hunk 2]
```diff
@@ -26,6 +26,33 @@
...
 from diffusion_planner.utils import ddp
+from planner_metrics.temporal_stability import (
+    compute_curvature_rate_batch,
+    compute_mean_abs_jerk_batch,
+    compute_replan_consistency_batch,
+    inter_frame_transform,
+)
+
+
+def _predict_ego_for_temporal_metrics(model, inputs, args, device, delay=0):
```
Comment (go-sakayori) [inline diffusion_planner/diffusion_planner/validate_model.py]: This helper duplicates `validate_model`'s input prep verbatim (zeroed `sampled_trajectories`, delay tensor, `heading_to_cos_sin`, normalizer). Input prep has changed recently (#173, #174), so a future change applied only inside `validate_model` would silently leave the replan metrics running on differently-prepared inputs. Can `validate_model` and this path share one prep helper?

File: planner_metrics/temporal_stability.py (hunk 1 above)
Comment (Max-Bin) [reply planner_metrics/temporal_stability.py]: Thanks for pointing this out. This has been fixed in the latest version: `planner_metrics.temporal_stability` now imports and reuses `planner_metrics.geometry._build_sg_diff_kernel`, and the generated kernels are cached by `(window, poly, deriv, delta, device, dtype)`. So we no longer keep a separate SG kernel implementation here or run repeated `pinv` solves per validation batch.

File: diffusion_planner/diffusion_planner/validate_model.py (hunk 2 above)
Comment (Max-Bin) [reply diffusion_planner/diffusion_planner/validate_model.py]: Thanks, agreed. This has also been fixed in the latest version: validation input preparation is now centralized in `_prepare_validation_inputs()`, and the temporal stability prediction path uses the same preparation helper as the normal validation flow. This should prevent the temporal metrics from silently drifting when the main validation input prep changes in the future.

**Review by go-sakayori — APPROVED** [review]
Please solve the conflict

**Comment by Max-Bin** [comment]
@go-sakayori Thanks, solved!
